jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/ExtractTarStep.java

```diff
@@ -146,42 +151,39 @@ public LocalImage call()
 
         for (int index = 0; index < layerFiles.size(); index++) {
           Path file = destination.resolve(layerFiles.get(index));
+          DescriptorDigest diffId = configurationTemplate.getLayerDiffId(index);
 
-          // Compress layers if necessary and calculate the digest/size
-          Blob blob = Blobs.from(file);
           try (ProgressEventDispatcher childDispatcher =
                   childProgressFactories
                       .get(index)
                       .create("compressing " + file, Files.size(file));
               ThrottledAccumulatingConsumer throttledProgressReporter =
                   new ThrottledAccumulatingConsumer(childDispatcher::dispatchProgress)) {
-            if (!layersAreCompressed) {
-              Path compressedFile = destination.resolve(layerFiles.get(index) + ".compressed");
-              try (GZIPOutputStream compressorStream =
-                      new GZIPOutputStream(Files.newOutputStream(compressedFile));
-                  NotifyingOutputStream notifyingOutputStream =
-                      new NotifyingOutputStream(compressorStream, throttledProgressReporter)) {
-                blob.writeTo(notifyingOutputStream);
+            Optional<CachedLayer> optionalLayer = cache.retrieveTarLayer(diffId);
+            if (optionalLayer.isPresent()) {
+              // Retrieve pre-compressed layer from cache
+              CachedLayer layer = optionalLayer.get();
+              layers.add(new PreparedLayer.Builder(layer).build());
+              v22Manifest.addLayer(layer.getSize(), layer.getDigest());
+            } else {
+              // Compress layers and calculate the digest/size
+              Blob blob = Blobs.from(file);
```

Comment:
Now that the loop has become pretty long, I think it warrants to rename `file` to `layerFile`. And the comment "calculate the digest/size" doesn't seem to hold anymore.